File: check_xups_alarms.pl
#!/usr/bin/perl -w

use strict;
use SNMP;
use Net::SNMP qw(SNMP_VERSION_1);
use Getopt::Std;

use constant MIB_UPS_ALARMS => '1.3.6.1.4.1.534.1.7';

use constant EXAMPLE => "\n\n".
    "Example:\n".
    "\n".
    "./check_xups_alarms.pl -H 192.168.0.1 -w 1..4,11 -c 5..10 -C pass\n".
    "\n".
    "It checks host  192.168.0.1 (UPS-MIB SNMP compliant device) looking for any active alarm\n".
    "present in both warning and critical lists.\n".
    "Plugin returns CRITICAL one or more alarms with id 5 to 10 are active,\n".
    "and WARNING if one or more alarms with id 1, 2, 3, 4 or 11 are active.\n".
    "In both two cases a list of active alarm ids and descriptions is returned.\n".
    "In other case it returns OK if check has been successfully performed.";

my %options=();
getopts("H:C:p:t:w:c:hu", \%options);

my $Result;
my $Output;

# Nagios exit codes
my $OKAY        = 0;
my $WARNING     = 1;
my $CRITICAL    = 2;
my $UNKNOWN     = 3;

# Command arguments and defaults
my $snmp_host           = $options{H};
my $snmp_community      = $options{C};
my $snmp_port           = $options{p} || 161;   # SNMP port default is 161
my $connection_timeout  = $options{t} || 10;    # Connection timeout default 10s
my $default_error       = (!defined $options{u}) ? $CRITICAL : $UNKNOWN;
my $high_precision      = (defined $options{h}) ? 1 : 0;
my $critical            = $options{c};
my $warning             = $options{w};
my $session;
my $error;
my $exitCode;


# APCs have a maximum length of 15 characters for snmp community strings
if(defined $snmp_community) {$snmp_community = substr($snmp_community,0,15);}

# If we don't have the needed command line arguments exit with UNKNOWN.
if(!defined $options{H} || !defined $options{C}){
	print "Not all required options were specified.\n\n";
    exit $UNKNOWN;
}

# Setup the SNMP session
($session, $error) = Net::SNMP->session(
    -hostname   => $snmp_host,
    -community  => $snmp_community,
    -timeout    => $connection_timeout,
    -port       => $snmp_port,
    -translate  => [-timeticks => 0x0]
);

# If we cannot build the SMTP session, error and exit
if (!defined $session) {
    my $output_header = ($default_error == $CRITICAL) ? "CRITICAL" : "UNKNOWN";
    printf "$output_header: %s\n", $error;
    exit $default_error;
}

if (&CheckArguments($error)) {
    # Argument checking passed
    $Result = &PerformCheck($Output);
	print "$Output \n";
	print "$Result \n";
	exit $OKAY;
}
else {
    # Error checking arguments
    $Output = $error;
    $Result = $UNKNOWN;
	print "$Output \n";
	print "$Result \n";
	exit $UNKNOWN;
}

sub CheckArguments() {
    my @IfRange;
    my $ArgOK;
    my $ThresholdsFormat;
    my $WarningAlarm;
    my @WARange;
    my @WarningAlarms;
    my $WARangeLength;
    my $CriticalAlarm;
    my @CARange;
    my @CriticalAlarms;
    my $CARangeLength;

    #Check Warnings Alarms Number
    if ($warning =~ /^(\d+|\d+\.\.\d+)(,(\d+|\d+\.\.\d+))*$/) { #One or more digits
        @WarningAlarms = split(/,/, $warning);
        foreach $WarningAlarm(@WarningAlarms) {
            @WARange = split(/\.\./, $WarningAlarm);
            $WARangeLength = @WARange;
            if ($WARangeLength > 1){ #It is an alarm range
                if ($WARange[0] >= $WARange[1]){
                    $_[0] = "Invalid warning alarm range.The first number must be lower than the second one.";
                    return 0;
                }
                else{
                    if ($WARange[0] < 0 || $WARange[0] > 65535 || $WARange[1] < 0 || $WARange[1] > 65535 ) {
                        $_[0] = "Invalid warning alarm number";
                        return 0;
                    }
                }
            }
            else{ #It is an Alarm Id
                if ($WarningAlarm < 0 || $WarningAlarm > 65535) {
                    $_[0] = "Invalid warning alarm number";
                    return 0;
                }
            }
        }
    }
    else{
        $_[0] = "Invalid warning alarm expression";
        return 0;
    }

    #Check Critical Alarms Number
    if ($critical =~ /^(\d+|\d+\.\.\d+)(,(\d+|\d+\.\.\d+))*$/) { #One or more digits
        @CriticalAlarms = split(/,/, $critical);
        foreach $CriticalAlarm(@CriticalAlarms) {
            @CARange = split(/\.\./, $CriticalAlarm);
            $CARangeLength = @CARange;
            if ($CARangeLength > 1){ # It is an alarm range
                if ($CARange[0] >= $CARange[1]){
                    $_[0] = "Invalid critical alarm range.The first number must be lower than the second one.";
                    return 0;
                }
                else{
                    if ($CARange[0] < 0 || $CARange[0] > 65535 || $CARange[1] < 0 || $CARange[1] > 65535 ) {
                        $_[0] = "Invalid critical alarm number";
                        return 0;
                    }
                }
            }
            else{ # It is an Alarm Id
                if ($CriticalAlarm < 0 || $CriticalAlarm > 65535) {
                    $_[0] = "Invalid critical alarm number";
                    return 0;
                }
            }
        }
    }
    else{
        $_[0] = "Invalid critical alarm expression";
        return 0;
    }

    return 1;
}


# Performs whole check:
# Input: Nagios Plugin object
# Output: Plugin output string
# Return value: Plugin return value

sub PerformCheck() {
    my $OID_UpsAlarmsPresent =   MIB_UPS_ALARMS . '.1.0';
    my $OID_UpsAlarmsTable =   MIB_UPS_ALARMS . '.2';

    my $SNMPSession;
    my $SNMPError;
    my @descriptionOId;
    my $alarmId;

    my $PluginOutput;
    my $PluginReturnValue = $UNKNOWN;
    my @RangeWarningAlarms;
    my @RangeCriticalAlarms;

    ($SNMPSession, $SNMPError) = Net::SNMP->session(
        -hostname   => $snmp_host,
        -community  => $snmp_community,
        -timeout    => $connection_timeout,
        -port       => $snmp_port,
    );

    if (!defined($SNMPSession)) {
        $PluginOutput = "Error '$SNMPError' starting SNMP session";
    }
    else {
        my $RequestResult = $SNMPSession->get_request(-varbindlist => [ $OID_UpsAlarmsPresent]);
        if (! defined $RequestResult) {
            # SNMP query error
            $PluginOutput = "Error '$SNMPError' retrieving info ".
                "from agent ${snmp_host}:${snmp_port} "
        }
        else{
            my $upsAlarmsPresent = $RequestResult->{$OID_UpsAlarmsPresent};
            if ($upsAlarmsPresent == 0){
                # If no alarms presents everything is ok and plugin finishes
                $PluginReturnValue = $OKAY;
                $PluginOutput = "No active alarms";
            }
            else {
                # One or more alarms are active
                my $WarningOutput = '';
                my $CriticalOutput = '';
                my @WarningAlarms = split(/,/, $warning);
                my @CriticalAlarms = split(/,/, $critical);
                my $AlarmsTable = $SNMPSession->get_table(-baseoid =>$OID_UpsAlarmsTable);
                #$Data::Dumper::Pair = " : ";
                #print Dumper($AlarmsTable);
                my $AlarmsActive="";
                my @AlarmMessages;
                my $description;

                #Extracted from XUPS-MIB
                $AlarmMessages[1]=" ";
                $AlarmMessages[2]=" ";
                $AlarmMessages[3]="UPS On Battery";
                $AlarmMessages[4]="LowBattery";
                $AlarmMessages[5]="UtilityPowerRestored";
                $AlarmMessages[6]="ReturnFromLowBattery";
                $AlarmMessages[7]="OutputOverload";
                $AlarmMessages[8]="Power Supply Fault";
                $AlarmMessages[9]="BatteryDischarged";
                $AlarmMessages[10]="InverterFailure";
                $AlarmMessages[11]="OnBypass";
                $AlarmMessages[12]="BypassNotAvailable";
                $AlarmMessages[13]="OutputOff";
                $AlarmMessages[14]="Input power Fault";
                $AlarmMessages[15]="BuildingAlarm";
                $AlarmMessages[16]="ShutdownImminent";
                $AlarmMessages[17]="OnInverter";
                $AlarmMessages[18]=" ";
                $AlarmMessages[19]=" ";
                $AlarmMessages[20]="BreakerOpen";
                $AlarmMessages[21]="AlarmEntryAdded";
                $AlarmMessages[22]="AlarmEntryRemoved";
                $AlarmMessages[23]="BatteryNeedService";
                $AlarmMessages[24]="OutputOffAsRequested";
                $AlarmMessages[25]="DiagnosticTestFailed";
                $AlarmMessages[26]="CommunicationsLost";
                $AlarmMessages[27]="UpsShutdownPending";
                $AlarmMessages[28]="AlarmTestInProgress";
                $AlarmMessages[29]="Temperature Fault";
                $AlarmMessages[30]="LossOfRedundancy";
                $AlarmMessages[31]="InternalTempBad";
                $AlarmMessages[32]="ChargerFailed";
                $AlarmMessages[33]="FanFailure";
                $AlarmMessages[34]="FuseFailure";
                $AlarmMessages[35]="PowerSwitchBad";
                $AlarmMessages[36]="ModuleFailure";
                $AlarmMessages[37]="OnAlternatePowerSource";
                $AlarmMessages[38]="AltPowerNotAvailable";
                $AlarmMessages[39]="UPS Fault";
                $AlarmMessages[40]="RemoteTempBad";
                $AlarmMessages[41]="RemoteHumidityBad";

                while(my ($k,$v)=each(%{$AlarmsTable})) {
                    if ($k =~ /1\.3\.6\.1\.4\.1\.534\.1\.7\.2\.1\.2/i) {
                        @descriptionOId = split(/\./, $$AlarmsTable{$k});

                        $alarmId = $descriptionOId[$#descriptionOId];
                        #if ($alarmId!="") {$AlarmsActive .= "$alarmId, ";}
                        $description="";
                        # Check if the alarm is a WellKnownAlarms to add description
                        if ($alarmId <= $#AlarmMessages){
                            $description = "$AlarmMessages[$alarmId]"
                        }
                        # Check if alarmId is one value of the criticalAlarm array or if it is in a range of the array
                        for (my $j=0; $j <=$#CriticalAlarms; $j++) {
                            @RangeCriticalAlarms = split(/\.\./, $CriticalAlarms[$j]);
                            if( $#RangeCriticalAlarms ) {
                                # Checking range
                                if( $alarmId >= $RangeCriticalAlarms[0] && $alarmId <= $RangeCriticalAlarms[1] ) {
                                    $CriticalOutput .= "$description ";
                                }
                            }
                            else{
                                if ( $alarmId == $RangeCriticalAlarms[0] ){
                                    $CriticalOutput .= "$description ";
                                }
                            }
                        }

                        if ($CriticalOutput eq '') {
                            # No critical alarms present, search warning alarms
                            for (my $j=0; $j <=$#WarningAlarms; $j++) {
                                @RangeWarningAlarms = split(/\.\./, $WarningAlarms[$j]);
                                if( $#RangeWarningAlarms ) {
                                    #Checking range
                                    if( $alarmId >= $RangeWarningAlarms[0] && $alarmId <=$RangeWarningAlarms[1] ) {
                                        $WarningOutput .= "$description ";
                                    }
                                }
                                else {
                                    if ($alarmId == $RangeWarningAlarms[0] ) {
                                        $WarningOutput .= "$description ";
                                    }
                                }
                            }
                        }
                    }
                }
                if ( $CriticalOutput ne '' ) {
                    $PluginReturnValue = $CRITICAL;
                    $PluginOutput = $CriticalOutput;
                }
                elsif ( $WarningOutput ne '' ) {
                    $PluginReturnValue = $WARNING;
                    $PluginOutput = $WarningOutput;
                }
                else {
                    $PluginReturnValue = $OKAY;
                    substr($AlarmsActive,-2)='';#erases last comma and blank
                    $AlarmsActive="(".$AlarmsActive.")";
                    $PluginOutput = "No active Alarms";
                    # $PluginOutput = "Alarms active $AlarmsActive"." but not set in check lists";
                }
            }
        }

        # Close SNMP session
        $SNMPSession->close();
    }

    #Return result
    $_[0] = $PluginOutput;
    return $PluginReturnValue;
}




sub query_oid {
    # This function will poll the active SNMP session and return the value
    # of the OID specified. Only inputs are OID. Will use global $session
    # variable for the session.
    my $oid = $_[0];
    my $response = $session->get_request(-varbindlist => [ $oid ],);

    # If there was a problem querying the OID error out and exit
    if (!defined $response) {
        my $output_header = ($default_error == $CRITICAL) ? "CRITICAL" : "UNKNOWN";
        printf "$output_header: %s\n", $session->error();
        $session->close();
        exit $default_error;
    }

    return $response->{$oid};
}

# The end. We shouldn't get here, but in case we do exit unknown
print "UNKNOWN: Unknown script error\n";
exit $UNKNOWN;

File: check_xups_alarms.py
#! /usr/bin/python

from pysnmp import hlapi
import getopt
import sys
import re

# Define oids
MIB_UPS_ALARMS = "1.3.6.1.4.1.534.1.7"


class CheckXUPS:
    help = """ Usage:
            -H  Address of hostname of UPS (required)
            -C  SNMP community string (required)
            -l  Command (optional, see command list)
            -p  SNMP port (optional, defaults to port 161)
            -c critical alarms
            -w warning alarms

            Commands (supplied with -l argument):
            
                manufacturer
                    manufacturer details

            Example:
            ./check_xups_alarms.py -H 10.xxx.xxxx.xx -C pass -w 1..4,11 -c 5..10
            """
    OKAY = 0
    WARNING = 1
    CRITICAL = 2
    UNKNOWN = 3

    def __init__(self):
        self.snmp_host = None
        self.snmp_community = None
        self.snmp_port = None
        self.warnings = None
        self.critical = None

        self.exitCode = None
        self.outputValue = None

        self.setup()

    def setup(self):
        argv = sys.argv[1:]
        # print(argv)

        try:
            opts, args = getopt.getopt(argv, "H:C:l:p:t:w:c:hu")
        except getopt.GetoptError as err:
            print(str(err).upper())
            return

        for opt, arg in opts:
            if opt in ['-H']:
                self.snmp_host = arg
            elif opt in ['-C']:
                self.snmp_community = hlapi.CommunityData(arg)
            elif opt in ['-p']:
                self.snmp_port = arg
            elif opt in ['-c']:
                self.critical = arg
            elif opt in ['-w']:
                self.warnings = arg

        if self.snmp_host is None or self.snmp_community is None:
            print("snmp_host -H or snmp_community -C is not defined")
            exit(1)

        if self.warnings is None or self.critical is None:
            print("alarms are not defined")
            print(self.help)
            exit(2)

        self.execute()

    def execute(self):

        if self.checkArguments():

            self.performCheck()

            if self.exitCode == self.CRITICAL:
                print("CRITICAL")
                # Do something
            elif self.exitCode == self.WARNING:
                print("WARNING")
                # Do something
            elif self.exitCode == self.OKAY:
                print("OKAY")
                # Nothing to do

            print(self.outputValue)
            return

    def checkArguments(self):
        if re.findall('^(\d+|\d+\.\.\d+)(,(\d+|\d+\.\.\d+))*$', self.warnings):
            WarningAlarms = re.split(',', self.warnings)
            self.warnings = []
            for element in WarningAlarms:
                WARange = re.split('\.\.', element)
                WARange = [int(numeric_string) for numeric_string in WARange]  # Convert to integers
                if len(WARange) > 1:  # It is an alarm range
                    if WARange[0] > WARange[1]:
                        print("Invalid warning alarm range.The first number must be lower than the second one.")
                        return 0
                    else:
                        if WARange[0] < 0 or WARange[0] > 65535 or WARange[1] < 0 or WARange[1] > 65535:
                            print("Invalid warning alarm number")
                            return 0
                    # Dobar unos
                    for x in range(WARange[0], WARange[1]):
                        self.warnings.append(x)
                else:  # It is an AlarmId
                    element = int(element)  # Convert to integer
                    if element < 0 or element > 65535:
                        print("Invalid warning alarm number")
                        return 0
                    self.warnings.append(element)

            if re.findall('^(\d+|\d+\.\.\d+)(,(\d+|\d+\.\.\d+))*$', self.critical):
                CriticalAlarms = re.split(',', self.critical)
                self.critical = []
                for element in CriticalAlarms:
                    CRange = re.split('\.\.', element)
                    CRange = [int(numeric_string) for numeric_string in CRange]  # Convert to integers
                    if len(CRange) > 1:  # It is an alarm range
                        if CRange[0] > CRange[1]:
                            print("Invalid warning alarm range.The first number must be lower than the second one.")
                            return 0
                        else:
                            if CRange[0] < 0 or CRange[0] > 65535 or CRange[1] < 0 or CRange[1] > 65535:
                                print("Invalid warning alarm number")
                                return 0
                        # Dobar unos
                        for x in range(CRange[0], CRange[1]):
                            self.critical.append(x)
                    else:  # It is an AlarmId
                        element = int(element)  # Convert to integer
                        if element < 0 or element > 65535:
                            print("Invalid warning alarm number")
                            return 0
                        self.critical.append(element)
        return 1

    def performCheck(self):
        OID_UpsAlarmsPresent = MIB_UPS_ALARMS + '.1.0'
        OID_UpsAlarmsTable = MIB_UPS_ALARMS + '.2'

        # Vraca dictionary OID: broj alarma
        RequestResult = self.get(self.snmp_host, [OID_UpsAlarmsPresent], self.snmp_community)

        presentAlarm = RequestResult[OID_UpsAlarmsPresent]

        if presentAlarm == 0:
            #If no alarms presents everything is ok and plugin finishes
            self.exitCode = self.OKAY
            self.outputValue = "No active alarms"
            return
        else:
            # Extracted from XUPS-MIB
            AlarmMessagesDitc = {1: "", 2: "", 3: "UPS On Battery", 4: "LowBattery", 5: "UtilityPowerRestored",
                                 6: "ReturnFromLowBattery", 7: "OutputOverload", 8: "Power Supply Fault",
                                 9: "BatteryDischarged", 10: "InverterFailure", 11: "OnBypass",
                                 12: "BypassNotAvailable", 13: "OutputOff", 14: "Input power Fault",
                                 15: "BuildingAlarm", 16: "ShutdownImminent", 17: "OnInverter", 18: "", 19: "",
                                 20: "BreakerOpen", 21: "AlarmEntryAdded", 22: "AlarmEntryRemoved",
                                 23: "BatteryNeedService", 24: "OutputOffAsRequested", 25: "DiagnosticTestFailed",
                                 26: "CommunicationsLost", 27: "UpsShutdownPending", 28: "AlarmTestInProgress",
                                 29: "Temperature Fault", 30: "LossOfRedundancy", 31: "InternalTempBad",
                                 32: "ChargerFailed", 33: "FanFailure", 34: "FuseFailure", 35: "PowerSwitchBad",
                                 36: "ModuleFailure", 37: "OnAlternatePowerSource", 38: "AltPowerNotAvailable",
                                 39: "UPS Fault", 40: "RemoteTempBad", 41: "RemoteHumidityBad"}

            AlarmsTable = self.get(self.snmp_host, [OID_UpsAlarmsTable], self.snmp_community)
            foundWaring = 0
            foundCritical = 0
            description = ""

            for elem in AlarmsTable.keys():

                key = re.findall('1.3.6.1.4.1.534.1.7.2.1.(\d)', elem)
                key = 3
                if key: #Exists not null
                    key = int(key)

                    if key in self.critical:
                        desc = "CRITICAL ALARM: " + AlarmMessagesDitc[key] + "OID: " + elem + "\n"
                        description += desc
                        foundCritical = 1

                    elif key in self.warnings:
                        desc = "WARNING ALARM: " + AlarmMessagesDitc[key] + "OID: " + elem + "\n"
                        description += desc
                        foundWaring = 1

            if foundCritical == 1:
                self.exitCode = self.CRITICAL
            elif foundWaring == 1:
                self.exitCode = self.WARNING
            else:
                self.exitCode = self.OKAY
                description = "No active Alarms"

            self.outputValue = description

            return

    @staticmethod
    def construct_object_types(list_of_oids):
        object_types = []
        for oid in list_of_oids:
            object_types.append(hlapi.ObjectType(hlapi.ObjectIdentity(oid)))
        return object_types

    @staticmethod
    def cast(value):
        try:
            return int(value)
        except (ValueError, TypeError):
            try:
                return float(value)
            except (ValueError, TypeError):
                try:
                    return str(value)
                except (ValueError, TypeError):
                    pass
        return value

    def fetch(self, handler, count):
        result = []
        for i in range(count):
            try:
                error_indication, error_status, error_index, var_binds = next(handler)
                if not error_indication and not error_status:
                    items = {}
                    for var_bind in var_binds:
                        items[str(var_bind[0])] = self.cast(var_bind[1])
                    result.append(items)
                else:
                    raise RuntimeError('Got SNMP error: {0}'.format(error_indication))
            except StopIteration:
                break
        return result

    def get(self, target, oids, credentials, port=161, engine=hlapi.SnmpEngine(), context=hlapi.ContextData()):
        handler = hlapi.getCmd(
            engine,
            credentials,
            hlapi.UdpTransportTarget((target, port)),
            context,
            *self.construct_object_types(oids)
        )
        return self.fetch(handler, 1)[0]


def main():
    apc = CheckXUPS()


if __name__ == "__main__":
    main()
